### PR TITLE
7903519 : jtreg/jtharness is missing features for basic crash testing

### PR DESCRIPTION
**Pull request by andrlos**
this does not exactly add a crash testing option, however it allows users to define their own resultModifiers that they can force jtharness to use via advertising their resultModifier service in META-INF. I tested it with basic crashOnlyModificator and it works as intended - allows me to correctly skip and pass all the regular failures and only leaves me with crash tests as failures. Perfect for corner cases like JVM testing and such.
Also checked correct behavior in the logs, where the regular failure is still being reported and only the justification of the result modification is added at the bottom.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Issue
 * [CODETOOLS-7903519](https://bugs.openjdk.org/browse/CODETOOLS-7903519): jtreg/jtharness is missing features for basic crash testing (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jtharness.git pull/57/head:pull/57` \
`$ git checkout pull/57`

Update a local copy of the PR: \
`$ git checkout pull/57` \
`$ git pull https://git.openjdk.org/jtharness.git pull/57/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 57`

View PR using the GUI difftool: \
`$ git pr show -t 57`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jtharness/pull/57.diff">https://git.openjdk.org/jtharness/pull/57.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jtharness/pull/57#issuecomment-1853025998)

### CODE REVIEW

File: src/com/sun/javatest/Script.java [hunk 1]
```diff
@@ -477,6 +481,57 @@ public void run() {
...
+    private void searchResultModifier(){
+        ServiceLoader<ResultModifierInterface> loader = ServiceLoader.load(ResultModifierInterface.class);
+
+        Iterator<ResultModifierInterface> iterator = loader.iterator();
+        ResultModifierInterface service = null;
+
+        while (iterator.hasNext()) {
+            if (service != null) {
+                // Found more than one implementation, throw an exception
+                throw new IllegalStateException("Multiple implementations found!");
```
Comment (judovana) [inline src/com/sun/javatest/Script.java:514]: Multiple implementations of what?

File: src/com/sun/javatest/Script.java [hunk 2]
```diff
@@ -477,6 +481,57 @@ public void run() {
...
 
+
+    /**
+     * This method runs the resultModifier stored in the private variable. In case the resultModifier is null
+     * it calls the method to search a resultModifierInterface implementation.
+     * @param proposedStatus - original status of the testrun
+     * @param td - description of the test
+     * @return status after modification (the proposedStatus by default)
+     */
+    private Status tryModifyResult(Status proposedStatus, TestDescription td){
```
Comment (judovana) [inline src/com/sun/javatest/Script.java:492]: this method should be synchronized, or the `resultModifier` should be safely  instantiated singleton. I would prefer the second. It would be good to die in initial stage, not late in runtime

File: src/com/sun/javatest/Script.java [hunk 3]
```diff
@@ -477,6 +481,57 @@ public void run() {
...
+        if (resultModifier == null) {
+            searchResultModifier();
+        }
+        return resultModifier.modifyStatus(proposedStatus, td);
```
Comment (judovana) [inline src/com/sun/javatest/Script.java:496]: maybe if a foreign code changed status result,  the FQN of class should be logged.

Comment (andrlos) [reply src/com/sun/javatest/Script.java:496]: but.. shouldnt user know this already, since he would be the one to put the foreign implementation there? I would leave it up to user to notify himself in the status line that the result has been modified.. alternative would be to check the status after calling "tryModifyResult" and if its not the same status, appending String saying that the Status has been modified by external code.. :shrug:

Comment (judovana) [reply src/com/sun/javatest/Script.java:496]: not necessarily. the one who set-up jtreg runner can be somebody else then who read the logs. Generally any jenkins infra - one person prepare generic runners, another jenkins jobs, and thne developer submit jdk, and read the results

File: src/com/sun/javatest/Script.java (hunk 2 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:492]: I prefer the method to be synchronized to be honest. I am not seeing how to keep the change this simple with mixing singleton in it..

Comment (andrlos) [reply src/com/sun/javatest/Script.java:492]: "private synchronized void searchResultModifier()" should do it, shouldnt it? this function is only called once per Script instance which I assume is once per run.

File: src/com/sun/javatest/Script.java (hunk 3 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:496]: I still believe it would be up to the user who meddles in it to properly document it.. but I can definitely try to put in a mechanism that compares the statuses and in case of difference adds a line to the status line

File: src/com/sun/javatest/exec/ResultModifierInterface.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.sun.javatest.Status;
+import com.sun.javatest.TestDescription;
+
+/**
+ * Enables users to create their own result modifications for purposes of different testing strategies.
+ * This allows user to omit certain kinds of failures or pinpoint tests that passed even though the jvm
+ * crashed in the process without being intended to.
+ */
+public interface ResultModifierInterface {
```
Comment (vlsi) [inline src/com/sun/javatest/exec/ResultModifierInterface.java:37]:
```suggestion
public interface ResultModifier {
```

Should it be just `ResultModifier` rather than `ResultModifierInterface`?

File: src/com/sun/javatest/exec/ResultModifierInterface.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    Status modifyStatus(Status originalStatus, TestDescription td);
```
Comment (vlsi) [inline src/com/sun/javatest/exec/ResultModifierInterface.java:38]: The interface name does not seem to be related with the method it has.

a) If you want `ResultModifier` interface, then the method should probably be `void modify(TestResult result)` rather than passing "status and test description" separately
b) An alternative option could be `TestResultTransfomer extends Function<TestResult, TestResult>`

File: src/com/sun/javatest/Script.java (hunk 2 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:492]: okay, my apology judovana, seems like there is a new Script instance for every testrun. That means that a singleton approach will have to be done.. -,-

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 5 above)
Comment (judovana) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:38]: IIRC, the `originalStatus` on input is necessary, because the original idea @andrlos  had, that it was changing only failed tests. 
Otherwise I agree with your comments. ty!

Comment (vlsi) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:38]: The implementation can access the original status with `result.getStatus()`.

In other words, the framework should populate the result as usual with `result.setStatus(execStatus)`, and then it could call `resultModifier.modifyResult(result)`

Comment (judovana) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:38]: I see. ty!

Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:38]: once you call result.setStatus the status of the result becomes unmodifiable.. the workaround of this would be more painful and I dont see any benefit in it

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:37]: I am under impression that naming interface with suffix "Interface" is kinda a good practice..

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 5 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java:38]: and if you are thinking about it @vlsi - The status is unmodifiable by design. I still stand by my implementation as a concept and believe it is pretty elegant.. I shall address judovana's remarks and improve naming a bit as I know that is not something that is particularly clear

File: src/com/sun/javatest/Script.java (hunk 1 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java]: this has been addressed

File: src/com/sun/javatest/Script.java (hunk 2 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java]: okay, this has been addressed with static atribute storing the implementation in Script instance and a synchronized method for setting this atribute..

File: src/com/sun/javatest/Script.java (hunk 3 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java]: this has been addressed hopefully to your satisfaction

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: but if you insist on renaming it I have no problem with it :-)

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 5 above)
Comment (vlsi) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: if the status is unmodifiable, then `TestResultTransformer` with method `TestResult transform(TestResult)` would be better as it would cover more cases rather than changing status alone

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (vlsi) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: Neither jtreg nor openjdk use `...Interface` suffixes. Java naming convention does not suggest using it as well: https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

The naming of `ResultModifier` does not correspond to the actual intention of the interface: the implementation does not **modify** the result.

File: src/com/sun/javatest/Script.java [hunk 6]
```diff
@@ -477,6 +481,63 @@ public void run() {
...
+
+    /**
+     * This method runs the statusModifier stored in the private variable. In case the statusModifier is null
+     * it calls the method to search a statusModifierInterface implementation.
+     * @param originalStatus - original status of the testrun
+     * @param td - description of the test
+     * @return status after modification (the proposedStatus by default)
+     */
+    private Status tryModifyStatus(Status originalStatus, TestDescription td){
+        if (statusModifier == null) {
```
Comment (vlsi) [inline src/com/sun/javatest/Script.java:493]: `statusModifier` is not `volatile`, so reading the field without synchronization is unsafe

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 5 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: but TestResult is also unmodifiable once you set the status.. you would need to create a whole new TestResult and then link it and somehow chirurgically remove the old TestResult to not get a conflict.. and we dont need to cover more cases.. what would we even try to transform other than the status? We want to retain all the logs the test execution produced as I see no added value in modifying those. Giving user too much power can be dangerous.

File: src/com/sun/javatest/Script.java (hunk 6 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:493]: I fail to see your problem. `tryModifyStatus` tries to read the `statusModifier`, if its null it calls a `synchronized` function to search for usable `statusModifier` which is then set and then the field remains unchanged for the rest of the execution.

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: Okay, will be renamed as `StatusTransformator` ?

File: src/com/sun/javatest/Script.java (hunk 6 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:493]: would adding a `volatile` keyword to the atribute be sufficient? I am hesitant to make the tagged method synchronized as that would potentially create a bottleneck

Comment (vlsi) [reply src/com/sun/javatest/Script.java:493]: Please check https://shipilev.net/blog/2014/safe-public-construction/#_safe_publication

The code reads `statusModifier` without synchronization, and it might observe uninitialized object. In other words, a `non-null` value does not mean it was fully initialized.

Comment (andrlos) [reply src/com/sun/javatest/Script.java:493]: @vlsi it just occured to me that I can initialize the `statusModifier` during the initialization of the `Script` instance, therefore there should be no problem reading it from multiple threads.. WDYT? Sorry for throwing half witted ideas out here.. and I read the blog you shared I can see why you dont like the current approach and I agree it must be altered.

Comment (vlsi) [reply src/com/sun/javatest/Script.java:493]: Initializing it along with `Script` initialization sounds reasonable to me, and it might indeed simplify the code.
For instance, you could treat `null` implementation as a missing one, so you don not need to have "default implementation" in the code.

Comment (andrlos) [reply src/com/sun/javatest/Script.java]: okay, should be done and yeah you are right, in this simplified code there is no need for default implementation and it seems a lot nicer this way, thanks!

File: src/com/sun/javatest/Script.java [hunk 7]
```diff
@@ -477,6 +484,49 @@ public void run() {
...
+     * @return status after modification (the originalStatus by default)
+     */
+    private Status tryModifyStatus(Status originalStatus, TestDescription td){
+        if (statusTransformer == null) {
+            return originalStatus;
+        }
+        //in case the status has been modified by this function we want to log it for any user controlling the results.
+        Status newStatus = statusTransformer.transform(originalStatus, td);
+        newStatus = new Status(newStatus.getType(), newStatus.getReason() +
+                    " - The status of this test result has been modified by external code.");
```
Comment (vlsi) [inline src/com/sun/javatest/Script.java:502]: What do you think of mentioning the transformer name?

```suggestion
                    " - The status of this test result has been modified by " + statusTransfomer.getClass().getName());
```

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (andrlos) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: renamed as StatusTransformer.. Transformator has been evoking the electronic component name in my brain and did not seem correct.. also changer seems suitable, but that is up to you

File: src/com/sun/javatest/Script.java (hunk 7 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:502]: might be nice, will add that. Thanks for your suggestion :-)

File: src/com/sun/javatest/exec/ResultModifierInterface.java (hunk 4 above)
Comment (vlsi) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: `StatusTransformator` looks great to me

Comment (vlsi) [reply src/com/sun/javatest/exec/ResultModifierInterface.java]: `StatusTransformer` looks great to me

File: src/com/sun/javatest/Script.java [hunk 8]
```diff
@@ -208,6 +211,10 @@ public abstract class Script {
     private boolean jtrIfPassed =
             System.getProperty("javatest.script.jtrIfPassed", "true").equals("true");
 
+    /**
+     * tries to search for a statusTransformer, no status transformations will be done if the value is null
```
Comment (vlsi) [inline src/com/sun/javatest/Script.java:215]: Technically speaking, the field does not "try to search"

I would suggest something like

```suggestion
     * Contains {@link StatusTransformer} that transforms the test result status, or {@code null} if no transformer service was found.
```

File: src/com/sun/javatest/Script.java [hunk 9]
```diff
@@ -208,6 +211,10 @@ public abstract class Script {
...
+     */
+    private static final StatusTransformer statusTransformer = searchStatusTransformer();
```
Comment (vlsi) [inline src/com/sun/javatest/Script.java:217]: Typically, `static final` fields are named in `UPPER_CASE`

```suggestion
    private static final StatusTransformer STATUS_TRANSFORMER = loadStatusTransformer();
```

File: src/com/sun/javatest/Script.java [hunk 10]
```diff
@@ -477,6 +484,49 @@ public void run() {
...
+    private static synchronized StatusTransformer searchStatusTransformer(){
+        ServiceLoader<StatusTransformer> loader = ServiceLoader.load(StatusTransformer.class);
+
+        Iterator<StatusTransformer> iterator = loader.iterator();
+        StatusTransformer service = null;
+
+        while (iterator.hasNext()) {
+            if (service != null) {
+                // Found more than one implementation, throw an exception
+                throw new IllegalStateException("Multiple implementations of the StatusTransformer found!");
```
[the remark this answers is not in the document]
Comment (andrlos) [reply src/com/sun/javatest/Script.java:523]: Well yeah I thought about that, but I believe that during the lookup there is no guarantee about the order in which the transformers would come, right? Our usecase would be to use this with a simple transformer that would transform it depending whether a crash happened or not during the execution.. however in some of our tests the crash is a desired outcome and we observe the behavior of our jvm during the crash - whether it links to correct bug reporting channels upon crash etc. In our scenarios we want to mark only crashes as failures and regular failures without crashes as passes.. also passed test in which a crash occurred would be also passed. I believe that there should only be one implementation allowed for now unless someone misses multiple implementations direly in the future, but that person can also squeeze implementation of multiple transformers into a single transformer, right?

I am not purely against, I am just not sure whether the benefits of allowing multiple implementations outweigh the possible errors due to forgotten implementations of transformers.

File: src/com/sun/javatest/Script.java (hunk 8 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:215]: sure, sorry comments and naming are not my strong suite

File: src/com/sun/javatest/Script.java (hunk 9 above)
Comment (andrlos) [reply src/com/sun/javatest/Script.java:217]: oh yeah.. forgot about that convention, will change :-)

File: src/com/sun/javatest/Script.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (vlsi) [reply src/com/sun/javatest/Script.java:525]: > but I believe that during the lookup there is no guarantee about the order in which the transformers would come, right? 

I believe there's a certain order that is related to classpath order, modulepath order.

[the remark this answers is not in the document]
Comment (judovana) [reply src/com/sun/javatest/Script.java:525]: I like the idea of chained transformers.  The only issue indeed is order. Our usecase is jsut one of the many.  And chaining gives interesting posibilitues.

Maybe the interface can have `int getOrder` method and implements comparable with default method comparing by this, so order would be deterministc, and mainly - in user's hand? 

in addition it will solve the issue with default impl x null check. The transfomers would be simply list, and if empty, none will be applied.

[the remark this answers is not in the document]
Comment (judovana) [reply src/com/sun/javatest/Script.java:525]: my trust to cp/mp order is a bit limited :) But may be much better then my artifical getOrder with sort

[the remark this answers is not in the document]
Comment (vlsi) [reply src/com/sun/javatest/Script.java:525]: Frankly, `getOrder` could probably be added later if needed (it could be added as a default implementation returning 0 or something).

So I agree `getOrder` might be helpful if exact ordering will be needed in the future, however, I guess we could skip adding `getOrder` this time and rely on `ServiceLoader` order

[the remark this answers is not in the document]
Comment (judovana) [reply src/com/sun/javatest/Script.java:525]: sure. ty!

**Comment by judovana** [comment]
@vlsi  I'm partially surprised that you seems to ok to include this change to jtharness. When @andrlos come with the idea I was a bit sceptic.  Now I'm curious, do you think jtreg project could harbour some implementations of this interface, and eg produce them as separate jars?

File: src/com/sun/javatest/Script.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (andrlos) [reply src/com/sun/javatest/Script.java:525]: So should I add it? Reading your covnersation I assume that yes. :smile:

**Comment by vlsi** [comment]
> I'm partially surprised that you seems to ok to include this change to jtharness

Frankly, I am still a bit skeptical, however, the new naming of `StatusTransformer + transform` sounds more-or-less good, and it seems to be the least evil when compared with fully replacing `TestResult` within `Script`.

The suggested PR does not add much maintenance overhead, however, it would probably make sense to add at least one test for it. @andrlos , have you considered adding a test for the feature? Frankly, I have no idea if it would be easy or hard to add a test (I've no idea how tests for jtharness are created), however, it would be great to have one.

Frankly, from a test framework user's point of view, I would prefer something like `TestResult transform(TestResult)` or `testResult.withStatus(newStatus) -> producing new TestResult` or `TestResult.setStatus(newStatus)`.
However, the latter suggestions might require significant rework of jtreg. Of course, I am nowhere near to asking for such changes.

----

> Now I'm curious, do you think jtreg project could harbour some implementations of this interface

I am not sure. I know what jtreg is, however, I virtually do not use it, so I can't tell.

**Comment by andrlos** [comment]
@vlsi I can try to add a test for this.. I will look into some of the tests already existing in the project and will see how hard it would be to add one..

**Comment by andrlos** [comment]
@vlsi so I am looking at the possible ways to implement a unit test for this behavior.. there are no tests for the Script.java class yet.. the problem there is that overly large method `run` calls half the functionality of jtharness before it even gets to the code addition.. I will keep thinking about it.. I know I can mock some behavior but this would be a lot. :smile: we will see

**Comment by andrlos** [comment]
@vlsi I am struggling running the unit tests via the ant target.. either I am doing it completely wrong or someone did not try to run them for a VERY long time, the amount of errors I get is astounding.. I dont suppose that the pipeline here on github is running any kind of testing that would be invisible to me as an outsider?

**Comment by judovana** [comment]
> @vlsi I am struggling running the unit tests via the ant target.. either I am doing it completely wrong or anyone did not try to run them for a VERY long time, the amount of errors I get is astounding.. I dont suppose that the pipeline here on github is running any kind of testing that would be invisible to me as an outsider?

Afaik adoptium is running build of jtregs: https://ci.adoptium.net/view/Dependencies/job/dependency_pipeline/

it builds jtregs and also run tests.
HAve you tried the make ? It is a bit better maintianed I think. Also the https://github.com/adoptium/ci-jenkins-pipelines/blob/master/tools/code-tools/jtreg.sh is using make.

**Comment by judovana** [comment]
hmm, looks like the tests are not run :(

**Comment by andrlos** [comment]
@vlsi I was able to make the unit test run locally in the end, there was an issue with my machine configuration that caused the errors.. I implemented simple unit test for Script behavior, however it does not test the usage of the services.. to test it completely I would have to modify the jar midexecution with new custom META-INF and it somehow feels wrong hardcoding the relative path to the javatest.jar into the unit test itself.. the other idea that came to mind was to just not test the service lookup but just the implementation usage, however for this I would need to make the `loadStatusTransformer` function not `static` and `protected` instead of `private`. WDYT?

File: unit-tests/com/sun/javatest/ScriptTest.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        for(String s: entries){
+            File currentFile = new File(file.getPath(),s);
+            if(currentFile.isDirectory()){
+                success = success && cleanup(currentFile);
+            }
+            else {
+                success = success && currentFile.delete();
+            }
+        }
+        return success && file.delete();
```
Comment (vlsi) [inline unit-tests/com/sun/javatest/ScriptTest.java:76]: I would suggest using JUnit's [TemporaryFolder](https://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html) or an alternative solution if it is already present in jtharness. Rolling "create and delete tempdir" has issues when it comes to different operating systems (e.g. Windows can't delete directory if there's open file), and it has security issues (e.g. someone might spy on the created files or something like that).

Comment (andrlos) [reply unit-tests/com/sun/javatest/ScriptTest.java:76]: @vlsi and what are your other thoughts about the test? is it enough? or should I go with the approach to make the loadTransformer method non static and protected so I can override it in the test or should I stick with this test and we can call it a day?

Comment (andrlos) [reply unit-tests/com/sun/javatest/ScriptTest.java]: I added the TemporaryFolder part and got rid of the cleanup method

**Comment by andrlos** [comment]
@dbessono first of all - thank you for your kind input. 
Second - in my experience with both jtharness and its integration into jtreg I don't think that the case you are trying to imply is as simple as you might think it is. There are very cryptic ways of how the TestResult is being propagated down and back up and initial testing of your idea gave me NPE as the TestResult gets lost if I try to meddle with those two methods alone and I believe it has to do something with the test internal urls and the shrinklist... I will see how deep this rabbit hole gets me.
Third - as we would want to have the option to modify the input build into the either jtreg or jtharness directly, is there any real issue of integrating this option directly to jtharness with this very simple piece of code? as the only advantage of using your proposed solution would be to keep jtharness code from changing and forcing us to do changes in the jtreg code.
Again thank you very much for your input.

**Comment by dbessono** [comment]
> @dbessono first of all - thank you for your kind input. Second - in my experience with both jtharness and its integration into jtreg I don't think that the case you are trying to imply is as simple as one might think it is. There are very cryptic ways of how the TestResult is being propagated down and back up and initial testing of your idea gave me NPE as the TestResult gets lost if I try to meddle with those two methods alone and I believe it has to do something with the test internal urls and the shrinklist... I will see how deep this rabbit hole gets me. Third - as we would want to have the option to modify the input built directly into the either jtreg or jtharness, is there any real issue of integrating this option directly to jtharness with this very simple piece of code? as the only advantage of using your proposed solution would be to keep jtharness code from changing and forcing us to do the changes in the jtreg code instead. Again thank you very much for your input.

Thanks for suggesting the customization.

Ah, can see that the `testResult` field is private. The original idea of `Script` base class is to allow users to subclass it and override the needed parts. So to avoid introduction of new concepts, interfaces etc. and to make customization simpler we can allow subclasses of `Script` to partially override the stages test execution flow (instead of forcing to completely reimplement method `::run`) by extracting phases of `::run` that could be overridden if needed, say smth like:

```
diff --git a/src/com/sun/javatest/Script.java b/src/com/sun/javatest/Script.java
--- a/src/com/sun/javatest/Script.java	(revision 22e872054eec91c94da79f8358cb010a33788755)
+++ b/src/com/sun/javatest/Script.java	(date 1704717221851)
@@ -465,7 +465,7 @@
 
         testResult.setEnvironment(env);
         testResult.putProperty("totalTime", Long.toString(System.currentTimeMillis() - startMs));
-        testResult.setStatus(execStatus);
+        fillTestExecutionStatus(testResult, execStatus);
 
         try {
             if (execStatus.getType() != Status.PASSED || jtrIfPassed) {
@@ -477,6 +477,13 @@
         }
     }
 
+    /**
+     * TODO
+     */
+    protected void fillTestExecutionStatus(TestResult testResultToFill, Status execStatus) {
+        testResultToFill.setStatus(execStatus);
+    }
+
     /**
      * The primary method to be provided by Scripts. It is responsible for compiling
      * and executing the test appropriately.  Normally, a script should call `init' and

```

Note that `protected TestDescription td;` field is accessible to the overriders.

**Comment by andrlos** [comment]
@dbessono sounds reasonable.. Will try that locally.

**Comment by andrlos** [comment]
@dbessono feast your eyes on your proposed approach.. hopefully we get to merge this since it was your idea anyway https://github.com/openjdk/jtharness/pull/59 I tested it with the thing I intend to do (the crashonly jtreg testing) and it works
